src/test/test-copy.c

```diff
@@ -50,6 +51,123 @@ TEST(copy_file) {
+        assert_se(write_string_file(dest, "foo foo foo", WRITE_STRING_FILE_CREATE) == 0);
+
+        assert_se(read_full_file(dest, &buf, &sz) == 0);
+        assert_se(streq(buf, "foo foo foo\n"));
+        assert_se(sz == 12);
+
+        /* The file exists- now overwite original contents, and test the COPY_REPLACE flag. */
+
+        assert_se(copy_tree(src, dest, UID_INVALID, GID_INVALID, COPY_REFLINK) < 0);
+
+        assert_se(read_full_file(dest, &buf, &sz) == 0);
+        assert_se(streq(buf, "foo foo foo\n"));
+        assert_se(sz == 12);
+
+        assert_se(copy_tree(src, dest, UID_INVALID, GID_INVALID, COPY_REFLINK|COPY_REPLACE) == 0);
+
+        assert_se(read_full_file(dest, &buf, &sz) == 0);
+        assert_se(streq(buf, "bar bar\n"));
+        assert_se(sz == 8);
+}
+
+TEST(copy_tree_replace_dirs) {
+        _cleanup_free_ char *buf = NULL;
+        size_t sz;
+        _cleanup_free_ char *src_path1, *src_path2, *dst_path1, *dst_path2;
+        _cleanup_(rm_rf_physical_and_freep) char *src_directory = NULL;
+        _cleanup_(rm_rf_physical_and_freep) char *dst_directory = NULL;
+        char file1[] = "foo_file";
+        char file2[] = "bar_file";
+
+        /* Create the random source/destination directories */
+        assert_se(mkdtemp_malloc("/tmp/dirXXXXXX", &src_directory) >= 0);
+        assert_se(mkdtemp_malloc("/tmp/dirXXXXXX", &dst_directory) >= 0);
+
+        /* Construct the source/desitnation filepaths (should have different dir name, but same file names within) */
+        assert_se(src_path1 = path_join(src_directory, file1));
+        assert_se(src_path2 = path_join(src_directory, file2));
+        assert_se(dst_path1 = path_join(dst_directory, file1));
+        assert_se(dst_path2 = path_join(dst_directory, file2));
+
+        /* Populate some data to differentiate the files. */
+        assert_se(write_string_file(src_path1, "src file 1", WRITE_STRING_FILE_CREATE) == 0);
+        assert_se(write_string_file(src_path2, "src file 2", WRITE_STRING_FILE_CREATE) == 0);
+
+        assert_se(write_string_file(dst_path1, "dest file 1", WRITE_STRING_FILE_CREATE) == 0);
+        assert_se(write_string_file(dst_path2, "dest file 2", WRITE_STRING_FILE_CREATE) == 0);
+
+        {
+                assert_se(read_full_file(src_path1, &buf, &sz) == 0);
+                assert_se(streq(buf, "src file 1\n"));
+                assert_se(sz == 11);
+
+                assert_se(read_full_file(src_path2, &buf, &sz) == 0);
+                assert_se(streq(buf, "src file 2\n"));
+                assert_se(sz == 11);
+
+                assert_se(read_full_file(dst_path1, &buf, &sz) == 0);
+                assert_se(streq(buf, "dest file 1\n"));
+                assert_se(sz == 12);
+
+                assert_se(read_full_file(dst_path2, &buf, &sz) == 0);
+                assert_se(streq(buf, "dest file 2\n"));
+                assert_se(sz == 12);
+        }
```

Comment:
I'd drop these assertions, we're testing that `copy_tree()` works, not `write_string_file()`